File: waltid-applications/waltid-wallet-demo-compose/sharedLogic/src/commonMain/kotlin/id/walt/walletdemo/compose/logic/VerifierDisplayName.kt
package id.walt.walletdemo.compose.logic

val VerifierDetails.displayName: String
    get() = humanReadableVerifierName(name = name, clientId = clientId, responseUri = responseUri)

fun humanReadableVerifierName(name: String?, clientId: String?, responseUri: String? = null): String {
    name?.trim()?.takeIf { it.isNotEmpty() }?.let { return it }

    val response = responseUri?.trim().orEmpty()
    val parsedClientId = clientId?.let(VerifierClientIdentifier::parse)

    return parsedClientId?.host()
        ?: parsedClientId?.displayName()
        ?: NetworkOrigin.parse(response)?.host
        ?: "Unknown verifier"
}

private fun VerifierClientIdentifier.host(): String? =
    when (scheme) {
        VerifierClientIdentifier.Scheme.RedirectUri,
        VerifierClientIdentifier.Scheme.PreRegistered,
        VerifierClientIdentifier.Scheme.OpenIdFederation -> NetworkOrigin.parse(value)?.host
        else -> null
    }

private fun VerifierClientIdentifier.displayName(): String? {
    scheme.genericLabel?.let { return it }
    if (scheme == VerifierClientIdentifier.Scheme.X509SanDns) {
        return value.takeIf { it.isNotBlank() }
    }
    return rawValue.takeIf { scheme == VerifierClientIdentifier.Scheme.PreRegistered && it.length <= 48 }
}

File: waltid-applications/waltid-wallet-demo-ios/iosApp/iosApp/Display/VerifierDisplayName.swift
import Foundation

enum VerifierDisplayName {
    static func value(verifierName: String?, clientID: String?, responseURI: URL?) -> String {
        if let verifierName = verifierName?.trimmingCharacters(in: .whitespacesAndNewlines),
           !verifierName.isEmpty {
            return verifierName
        }

        if let clientIdentifier = clientID.flatMap(VerifierClientIdentifier.parse) {
            if let host = clientIdentifier.host {
                return host
            }
            if let displayName = clientIdentifier.displayName {
                return displayName
            }
        }

        if let host = responseURI?.host, !host.isEmpty {
            return host
        }

        return "Unknown verifier"
    }

}

private extension VerifierClientIdentifier {
    var host: String? {
        switch scheme {
        case .redirectURI, .preRegistered, .openIDFederation:
            return NetworkOrigin.parse(value)?.host
        case .x509SanDNS, .x509Hash, .decentralizedIdentifier, .legacyDID, .verifierAttestation, .unsupported:
            return nil
        }
    }

    var displayName: String? {
        if let genericLabel = scheme.genericLabel {
            return genericLabel
        }
        if scheme == .x509SanDNS {
            return value.isEmpty ? nil : value
        }
        if scheme == .preRegistered, rawValue.count <= 48 {
            return rawValue
        }
        return nil
    }
}
